File: stripe_terminal/ios/Classes/Mappings/ReaderMappings.swift
import Foundation
import StripeTerminal


extension Reader {
    func toApi() -> ReaderApi {
        return ReaderApi(
            availableUpdate: availableUpdate != nil,
            batteryLevel: batteryLevel?.doubleValue ?? -1.0,
            deviceType: deviceType.toApi(),
            label: label,
            location: location?.toApi(),
            locationId: locationId,
            locationStatus: locationStatus.toApi(),
            serialNumber: serialNumber,
            simulated: simulated
        )
    }
}

extension Location {
    func toApi() -> LocationApi {
        return LocationApi(
            address: address?.toApi(),
            displayName: displayName,
            id: stripeId,
            livemode: livemode,
            metadata: metadata ?? [:]
        )
    }
}


extension Address {
    func toApi() -> AddressApi? {
        return AddressApi(
            city: city,
            country: country,
            line1: line1,
            line2: line2,
            postalCode: postalCode,
            state: state
        )
    }
}


extension LocationStatus {
    func toApi() -> LocationStatusApi? {
        switch self {
        case .unknown:
            return nil
        case .set:
            return .set
        case .notSet:
            return .notSet
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension DeviceType {
    func toApi() -> DeviceTypeApi {
        switch self {
        case .chipper1X:
            return .chipper1X
        case .chipper2X:
            return .chipper2X
        case .stripeM2:
            return .stripeM2
        case .tapToPay:
            return .tapToPay
        case .verifoneP400:
            return .verifoneP400
        case .wisePad3:
            return .wisePad3
        case .wisePosE:
            return .wisePosE
        case .wisePosEDevKit:
            return .wisePosEDevkit
        case .etna:
            return .etna
        case .wiseCube:
            return .wiseCube
        case .stripeS700:
            return .stripeS700
        case .stripeS700DevKit:
            return .stripeS700Devkit
        case .verifoneV660p:
            return .verifoneV660p
        case .verifoneM425:
            return .verifoneM425
        case .verifoneM450:
            return .verifoneM450
        case .verifoneP630:
            return .verifoneP630
        case .verifoneUX700:
            return .verifoneUX700
        @unknown default:
            fatalError("DeviceType \(self) not supported.")
        }
    }
}

extension ReaderEvent {
    func toApi() -> ReaderEventApi {
        switch (self) {
        case .cardInserted:
            return .cardInserted
        case .cardRemoved:
            return .cardRemoved
        @unknown default:
            fatalError("ReaderEvent \(self) not supported.")
        }
    }
}

extension ReaderDisplayMessage {
    func toApi() -> ReaderDisplayMessageApi {
        switch (self) {
        case .retryCard:
            return .retryCard
        case .insertCard:
            return .insertCard
        case .insertOrSwipeCard:
            return .insertOrSwipeCard
        case .swipeCard:
            return .swipeCard
        case .removeCard:
            return .removeCard
        case .multipleContactlessCardsDetected:
            return .multipleContactlessCardsDetected
        case .tryAnotherReadMethod:
            return .tryAnotherReadMethod
        case .tryAnotherCard:
            return .tryAnotherCard
        case .cardRemovedTooEarly:
            return .cardRemovedTooEarly
        @unknown default:
            fatalError("ReaderDisplayMessage \(self) not supported.")
        }
    }
}

extension ReaderInputOptions {
    func toApi() -> [ReaderInputOptionApi] {
        var options: [ReaderInputOptionApi] = []
        if (contains(ReaderInputOptions.insertCard)) { options.append(ReaderInputOptionApi.insertCard) }
        if (contains(ReaderInputOptions.swipeCard)) { options.append(ReaderInputOptionApi.swipeCard) }
        if (contains(ReaderInputOptions.tapCard)) { options.append(ReaderInputOptionApi.tapCard) }
        return options
    }
}

extension BatteryStatus {
    func toApi() -> BatteryStatusApi? {
        switch (self) {
        case .critical:
            return .critical
        case .low:
            return .low
        case .nominal:
            return .nominal
        case .unknown:
            return nil
        @unknown default:
            fatalError("BatteryStatys \(self) not supported.")
        }
    }
}

extension ReaderSoftwareUpdate {
    func toApi() -> ReaderSoftwareUpdateApi {
        return ReaderSoftwareUpdateApi(
            components: components.toApi(),
            keyProfileName: nil,
            onlyInstallRequiredUpdates: false,
            requiredAt: requiredAt,
            settingsVersion: nil,
            timeEstimate: durationEstimate.toApi(),
            version: deviceSoftwareVersion
        )
    }
}

extension UpdateComponent {
    func toApi() -> [UpdateComponentApi] {
        var components: [UpdateComponentApi] = []
        if (contains(UpdateComponent.incremental)) {components.append(UpdateComponentApi.incremental)}
        if (contains(UpdateComponent.firmware)) {components.append(UpdateComponentApi.firmware)}
        if (contains(UpdateComponent.config)) {components.append(UpdateComponentApi.config)}
        if (contains(UpdateComponent.keys)) {components.append(UpdateComponentApi.keys)}
        return components
    }
}
    
extension UpdateTimeEstimate {
    func toApi() -> UpdateTimeEstimateApi {
        switch self {
        case .estimateLessThan1Minute:
            return .lessThanOneMinute
        case .estimate1To2Minutes:
            return .oneToTwoMinutes
        case .estimate2To5Minutes:
            return .twoToFiveMinutes
        case .estimate5To15Minutes:
            return .fiveToFifteenMinutes
        @unknown default:
            fatalError("UpdateTimeEstimate \(self) not supported.")
        }
    }
}

// PARAMS

extension DiscoveryConfigurationApi {
    func toHost() throws -> DiscoveryConfiguration? {
        switch self {
        case let config as BluetoothDiscoveryConfigurationApi:
            return try BluetoothScanDiscoveryConfigurationBuilder()
                .setTimeout(UInt(config.timeout ?? 0))
                .setSimulated(config.isSimulated)
                .build()
        case let config as BluetoothProximityDiscoveryConfigurationApi:
            return try BluetoothProximityDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .build()
        case _ as HandoffDiscoveryConfigurationApi:
            return nil
        case let config as InternetDiscoveryConfigurationApi:
            return try InternetDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .setLocationId(config.locationId)
                .build()
        case let config as TapToPayDiscoveryConfigurationApi:
            return try TapToPayDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .build()
        case _ as UsbDiscoveryConfigurationApi:
            return nil
        default:
            fatalError()
        }
    }
    
    func toHostDiscoveryMethod() -> DiscoveryMethod? {
        switch self {
        case _ as BluetoothDiscoveryConfigurationApi:
            return .bluetoothScan
        case _ as BluetoothProximityDiscoveryConfigurationApi:
            return .bluetoothProximity
        case _ as HandoffDiscoveryConfigurationApi:
            return nil
        case _ as InternetDiscoveryConfigurationApi:
            return .internet
        case _ as TapToPayDiscoveryConfigurationApi:
            return .tapToPay
        case _ as UsbDiscoveryConfigurationApi:
            return nil
        default:
            fatalError()
        }
    }
    
    func toHostSimulated() -> Bool {
        switch self {
        case let config as BluetoothDiscoveryConfigurationApi:
            return config.isSimulated
        case let config as BluetoothProximityDiscoveryConfigurationApi:
            return config.isSimulated
        case _ as HandoffDiscoveryConfigurationApi:
            return false
        case let config as InternetDiscoveryConfigurationApi:
            return config.isSimulated
        case let config as TapToPayDiscoveryConfigurationApi:
            return config.isSimulated
        case _ as UsbDiscoveryConfigurationApi:
            return false
        default:
            fatalError()
        }
    }
}

extension DeviceTypeApi {
    func toHost() -> DeviceType? {
        switch self {
        case .chipper2X:
            return .chipper2X
        case .verifoneP400:
            return .verifoneP400
        case .wisePad3:
            return .wisePad3
        case .stripeM2:
            return .stripeM2
        case .wisePosE:
            return .wisePosE
        case .wisePosEDevkit:
            return .wisePosEDevKit
        case .etna:
            return .etna
        case .chipper1X:
            return .chipper1X
        case .wiseCube:
            return .wiseCube
        case .stripeS700:
            return .stripeS700
        case .stripeS700Devkit:
            return .stripeS700DevKit
        case .tapToPay:
            return .tapToPay
        case .stripeS710, .stripeS710Devkit, .wisePad3s:
            return nil
        case .verifoneV660p:
            return .verifoneV660p
        case .verifoneM425:
            return .verifoneM425
        case .verifoneM450:
            return .verifoneM450
        case .verifoneP630:
            return .verifoneP630
        case .verifoneUX700:
            return .verifoneUX700
        case .verifoneV660pDevkit:
            return .verifoneV660pDevKit
        case .verifoneUX700Devkit:
            return .verifoneUX700DevKit
        @unknown default:
            fatalError("DeviceType \(self) not supported.")
        }
    }
}


extension CartApi {
    func toHost() throws -> Cart {
        return try CartBuilder(currency: currency)
            .setTax(tax)
            .setTotal(total)
            .setLineItems(lineItems.map { try $0.toHost()} )
            .build()
    }
}

extension CartLineItemApi {
    func toHost() throws -> CartLineItem {
        return try CartLineItemBuilder(displayName: description)
            .setAmount(amount)
            .setQuantity(quantity)
            .build()
    }
}

// EXTRA

extension ConnectionStatus {
    func toApi() -> ConnectionStatusApi {
        switch self {
        case .notConnected:
            return .notConnected
        case .discovering:
            return .discovering
        case .connecting:
            return .connecting
        case .connected:
            return .connected
        @unknown default:
            fatalError("WTF")
        }
    }
}

File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/ReaderMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.Address
import com.stripe.stripeterminal.external.models.BatteryStatus
import com.stripe.stripeterminal.external.models.Cart
import com.stripe.stripeterminal.external.models.CartLineItem
import com.stripe.stripeterminal.external.models.ConnectionStatus
import com.stripe.stripeterminal.external.models.DeviceType
import com.stripe.stripeterminal.external.models.DiscoveryConfiguration
import com.stripe.stripeterminal.external.models.Location
import com.stripe.stripeterminal.external.models.LocationStatus
import com.stripe.stripeterminal.external.models.Reader
import com.stripe.stripeterminal.external.models.ReaderDisplayMessage
import com.stripe.stripeterminal.external.models.ReaderEvent
import com.stripe.stripeterminal.external.models.ReaderInputOptions
import com.stripe.stripeterminal.external.models.ReaderSoftwareUpdate
import mek.stripeterminal.api.AddressApi
import mek.stripeterminal.api.BatteryStatusApi
import mek.stripeterminal.api.BluetoothDiscoveryConfigurationApi
import mek.stripeterminal.api.BluetoothProximityDiscoveryConfigurationApi
import mek.stripeterminal.api.CartApi
import mek.stripeterminal.api.CartLineItemApi
import mek.stripeterminal.api.ConnectionStatusApi
import mek.stripeterminal.api.DeviceTypeApi
import mek.stripeterminal.api.DiscoveryConfigurationApi
import mek.stripeterminal.api.HandoffDiscoveryConfigurationApi
import mek.stripeterminal.api.InternetDiscoveryConfigurationApi
import mek.stripeterminal.api.LocationApi
import mek.stripeterminal.api.LocationStatusApi
import mek.stripeterminal.api.ReaderApi
import mek.stripeterminal.api.ReaderDisplayMessageApi
import mek.stripeterminal.api.ReaderEventApi
import mek.stripeterminal.api.ReaderInputOptionApi
import mek.stripeterminal.api.ReaderSoftwareUpdateApi
import mek.stripeterminal.api.TapToPayDiscoveryConfigurationApi
import mek.stripeterminal.api.UpdateComponentApi
import mek.stripeterminal.api.UpdateTimeEstimateApi
import mek.stripeterminal.api.UsbDiscoveryConfigurationApi
import mek.stripeterminal.microsecondsToSeconds
import mek.stripeterminal.toHashMap

fun Reader.toApi(): ReaderApi {
    return ReaderApi(
        locationStatus = locationStatus.toApi(),
        batteryLevel = batteryLevel?.toDouble() ?: -1.0,
        deviceType = deviceType.toApi(),
        simulated = isSimulated,
        availableUpdate = availableUpdate?.hasFirmwareUpdate ?: false,
        locationId = location?.id,
        location = location?.toApi(),
        label = label,
        serialNumber = serialNumber!!
    )
}

fun LocationStatus.toApi(): LocationStatusApi? {
    return when (this) {
        LocationStatus.UNKNOWN -> null
        LocationStatus.SET -> LocationStatusApi.SET
        LocationStatus.NOT_SET -> LocationStatusApi.NOT_SET
    }
}

fun DeviceType.toApi(): DeviceTypeApi? {
    return when (this) {
        DeviceType.CHIPPER_1X -> DeviceTypeApi.CHIPPER1_X
        DeviceType.CHIPPER_2X -> DeviceTypeApi.CHIPPER2_X
        DeviceType.STRIPE_M2 -> DeviceTypeApi.STRIPE_M2
        DeviceType.TAP_TO_PAY_DEVICE -> DeviceTypeApi.TAP_TO_PAY
        DeviceType.VERIFONE_P400 -> DeviceTypeApi.VERIFONE_P400
        DeviceType.WISECUBE -> DeviceTypeApi.WISE_CUBE
        DeviceType.WISEPAD_3 -> DeviceTypeApi.WISE_PAD3
        DeviceType.WISEPAD_3S -> DeviceTypeApi.WISE_PAD3S
        DeviceType.WISEPOS_E -> DeviceTypeApi.WISE_POS_E
        DeviceType.WISEPOS_E_DEVKIT -> DeviceTypeApi.WISE_POS_E_DEVKIT
        DeviceType.ETNA -> DeviceTypeApi.ETNA
        DeviceType.STRIPE_S700 -> DeviceTypeApi.STRIPE_S700
        DeviceType.STRIPE_S700_DEVKIT -> DeviceTypeApi.STRIPE_S700_DEVKIT
        DeviceType.STRIPE_S710 -> DeviceTypeApi.STRIPE_S710
        DeviceType.STRIPE_S710_DEVKIT -> DeviceTypeApi.STRIPE_S710_DEVKIT
        DeviceType.UNKNOWN -> null
    }
}

fun Location.toApi(): LocationApi {
    return LocationApi(
        address = address?.toApi(),
        displayName = displayName,
        id = id,
        livemode = livemode,
        metadata = metadata?.toHashMap() ?: hashMapOf()
    )
}

fun Address.toApi(): AddressApi {
    return AddressApi(
        city = city,
        country = country,
        line1 = line1,
        line2 = line2,
        postalCode = postalCode,
        state = state
    )
}

fun ReaderEvent.toApi(): ReaderEventApi {
    return when (this) {
        ReaderEvent.CARD_INSERTED -> ReaderEventApi.CARD_INSERTED
        ReaderEvent.CARD_REMOVED -> ReaderEventApi.CARD_REMOVED
    }
}

fun ReaderDisplayMessage.toApi(): ReaderDisplayMessageApi {
    return when (this) {
        ReaderDisplayMessage.CHECK_MOBILE_DEVICE -> ReaderDisplayMessageApi.CHECK_MOBILE_DEVICE
        ReaderDisplayMessage.RETRY_CARD -> ReaderDisplayMessageApi.RETRY_CARD
        ReaderDisplayMessage.INSERT_CARD -> ReaderDisplayMessageApi.INSERT_CARD
        ReaderDisplayMessage.INSERT_OR_SWIPE_CARD -> ReaderDisplayMessageApi.INSERT_OR_SWIPE_CARD
        ReaderDisplayMessage.SWIPE_CARD -> ReaderDisplayMessageApi.SWIPE_CARD
        ReaderDisplayMessage.REMOVE_CARD -> ReaderDisplayMessageApi.REMOVE_CARD
        ReaderDisplayMessage.MULTIPLE_CONTACTLESS_CARDS_DETECTED ->
            ReaderDisplayMessageApi.MULTIPLE_CONTACTLESS_CARDS_DETECTED
        ReaderDisplayMessage.TRY_ANOTHER_READ_METHOD -> ReaderDisplayMessageApi.TRY_ANOTHER_READ_METHOD
        ReaderDisplayMessage.TRY_ANOTHER_CARD -> ReaderDisplayMessageApi.TRY_ANOTHER_CARD
        ReaderDisplayMessage.CARD_REMOVED_TOO_EARLY -> ReaderDisplayMessageApi.CARD_REMOVED_TOO_EARLY
    }
}

fun ReaderInputOptions.ReaderInputOption.toApi(): ReaderInputOptionApi? {
    return when (this) {
        ReaderInputOptions.ReaderInputOption.NONE -> null
        ReaderInputOptions.ReaderInputOption.INSERT -> ReaderInputOptionApi.INSERT_CARD
        ReaderInputOptions.ReaderInputOption.SWIPE -> ReaderInputOptionApi.SWIPE_CARD
        ReaderInputOptions.ReaderInputOption.TAP -> ReaderInputOptionApi.TAP_CARD
        ReaderInputOptions.ReaderInputOption.MANUAL_ENTRY -> ReaderInputOptionApi.MANUAL_ENTRY
    }
}

fun BatteryStatus.toApi(): BatteryStatusApi? {
    return when (this) {
        BatteryStatus.UNKNOWN -> null
        BatteryStatus.CRITICAL -> BatteryStatusApi.CRITICAL
        BatteryStatus.LOW -> BatteryStatusApi.LOW
        BatteryStatus.NOMINAL -> BatteryStatusApi.NOMINAL
    }
}

fun ReaderSoftwareUpdate.toApi(): ReaderSoftwareUpdateApi {
    return ReaderSoftwareUpdateApi(
        components = components.map { it.toApi() },
        keyProfileName = keyProfileName,
        onlyInstallRequiredUpdates = onlyInstallRequiredUpdates,
        requiredAt = requiredAtMs,
        settingsVersion = settingsVersion,
        timeEstimate = durationEstimate.toApi(),
        version = version
    )
}

fun ReaderSoftwareUpdate.UpdateComponent.toApi(): UpdateComponentApi {
    return when (this) {
        ReaderSoftwareUpdate.UpdateComponent.INCREMENTAL -> UpdateComponentApi.INCREMENTAL
        ReaderSoftwareUpdate.UpdateComponent.FIRMWARE -> UpdateComponentApi.FIRMWARE
        ReaderSoftwareUpdate.UpdateComponent.CONFIG -> UpdateComponentApi.CONFIG
        ReaderSoftwareUpdate.UpdateComponent.KEYS -> UpdateComponentApi.KEYS
    }
}

fun ReaderSoftwareUpdate.UpdateDurationEstimate.toApi(): UpdateTimeEstimateApi {
    return when (this) {
        ReaderSoftwareUpdate.UpdateDurationEstimate.LESS_THAN_ONE_MINUTE ->
            UpdateTimeEstimateApi.LESS_THAN_ONE_MINUTE
        ReaderSoftwareUpdate.UpdateDurationEstimate.ONE_TO_TWO_MINUTES ->
            UpdateTimeEstimateApi.ONE_TO_TWO_MINUTES
        ReaderSoftwareUpdate.UpdateDurationEstimate.TWO_TO_FIVE_MINUTES ->
            UpdateTimeEstimateApi.TWO_TO_FIVE_MINUTES
        ReaderSoftwareUpdate.UpdateDurationEstimate.FIVE_TO_FIFTEEN_MINUTES ->
            UpdateTimeEstimateApi.FIVE_TO_FIFTEEN_MINUTES
    }
}

// PARAMS

fun DiscoveryConfigurationApi.toHost(): DiscoveryConfiguration? {
    return when (this) {
        is BluetoothDiscoveryConfigurationApi ->
            DiscoveryConfiguration.BluetoothDiscoveryConfiguration(
                isSimulated = isSimulated,
                timeout = timeout?.let { microsecondsToSeconds(it) } ?: 0
            )
        is BluetoothProximityDiscoveryConfigurationApi -> null
        is HandoffDiscoveryConfigurationApi -> DiscoveryConfiguration.HandoffDiscoveryConfiguration()
        is InternetDiscoveryConfigurationApi ->
            DiscoveryConfiguration.InternetDiscoveryConfiguration(
                isSimulated = isSimulated,
                location = locationId,
                timeout = timeout?.let { microsecondsToSeconds(it) } ?: 0
            )
        is TapToPayDiscoveryConfigurationApi ->
            DiscoveryConfiguration.TapToPayDiscoveryConfiguration(
                isSimulated = isSimulated
            )
        is UsbDiscoveryConfigurationApi ->
            DiscoveryConfiguration.UsbDiscoveryConfiguration(
                isSimulated = isSimulated,
                timeout = timeout?.let { microsecondsToSeconds(it) } ?: 0
            )
    }
}

fun DeviceTypeApi.toHost(): DeviceType? {
    return when (this) {
        DeviceTypeApi.CHIPPER1_X -> DeviceType.CHIPPER_1X
        DeviceTypeApi.CHIPPER2_X -> DeviceType.CHIPPER_2X
        DeviceTypeApi.STRIPE_M2 -> DeviceType.STRIPE_M2
        DeviceTypeApi.TAP_TO_PAY -> DeviceType.TAP_TO_PAY_DEVICE
        DeviceTypeApi.VERIFONE_P400 -> DeviceType.VERIFONE_P400
        DeviceTypeApi.WISE_CUBE -> DeviceType.WISECUBE
        DeviceTypeApi.WISE_PAD3 -> DeviceType.WISEPAD_3
        DeviceTypeApi.WISE_POS_E -> DeviceType.WISEPOS_E
        DeviceTypeApi.WISE_PAD3S -> DeviceType.WISEPAD_3S
        DeviceTypeApi.WISE_POS_E_DEVKIT -> DeviceType.WISEPOS_E_DEVKIT
        DeviceTypeApi.ETNA -> DeviceType.ETNA
        DeviceTypeApi.STRIPE_S700 -> DeviceType.STRIPE_S700
        DeviceTypeApi.STRIPE_S700_DEVKIT -> DeviceType.STRIPE_S700_DEVKIT
        DeviceTypeApi.STRIPE_S710 -> DeviceType.STRIPE_S710
        DeviceTypeApi.STRIPE_S710_DEVKIT -> DeviceType.STRIPE_S710_DEVKIT
    }
}

fun CartApi.toHost(): Cart {
    return Cart.Builder(
        currency = currency,
        tax = tax,
        total = total,
        lineItems = lineItems.map { it.toHost() }
    )
        .build()
}

fun CartLineItemApi.toHost(): CartLineItem {
    return CartLineItem.Builder(
        description = description,
        quantity = quantity.toInt(),
        amount = amount
    )
        .build()
}

// EXTRA

fun ConnectionStatus.toApi(): ConnectionStatusApi {
    return when (this) {
        ConnectionStatus.NOT_CONNECTED -> ConnectionStatusApi.NOT_CONNECTED
        ConnectionStatus.CONNECTING -> ConnectionStatusApi.CONNECTING
        ConnectionStatus.CONNECTED -> ConnectionStatusApi.CONNECTED
        ConnectionStatus.DISCOVERING -> ConnectionStatusApi.DISCOVERING
    }
}
